google/resource_dns_record_set.go

```diff
@@ -55,6 +57,101 @@ func resourceDnsRecordSet() *schema.Resource {
 	}
 }
 
+// This catches most invalid DNS names, though it might not catch some invlaid non-ASCII names
+// it does not implement the TOASCII functionality defined in https://tools.ietf.org/html/rfc3454
+// See:
+// https://tools.ietf.org/html/rfc1034 DOMAIN NAMES - CONCEPTS AND FACILITIES
+// https://tools.ietf.org/html/rfc3490 Internationalizing Domain Names in Applications (IDNA)
+// https://tools.ietf.org/html/rfc3454 Preparation of Internationalized Strings ('stringprep')
+func validate_dns_rrset_name(name string) error {
+	if len(name) > 255 {
+		return fmt.Errorf("DNS name length %d greater than 254", len(name))
+	}
+
+	re, err := regexp.Compile("[\u002e\u3002\uFF0E\uFF61]$")
+	if err != nil {
+		return fmt.Errorf("Internal error compiling regexp: %s", err)
+	}
+	if !re.MatchString(name) {
+		return fmt.Errorf("DNS name must be fully qualified with a trailing dot")
+	}
+
+	dot_re, err := regexp.Compile("[\u002e\u3002\uFF0E\uFF61]")
+	if err != nil {
+		return fmt.Errorf("Internal error compiling regexp: %s", err)
+	}
+
+	verboten_re, err := regexp.Compile("[\x00-\x2C\x2E-\x2F\x3A-\x40\x5B-\x60\x7B-\x7F]+")
+	if err != nil {
+		return fmt.Errorf("Internal error compiling regexp: %s", err)
+	}
+
+	all_digit_re, err := regexp.Compile("^[[:digit:]]+$")
+	if err != nil {
+		return fmt.Errorf("Internal error compiling regexp: %s", err)
+	}
+
+	space_re, err := regexp.Compile("[[:space:]]+")
+	if err != nil {
+		return fmt.Errorf("Internal error compiling regexp: %s", err)
+	}
+
+	underscore_re, err := regexp.Compile("_+")
+	if err != nil {
+		return fmt.Errorf("Internal error compiling regexp: %s", err)
+	}
+
+	labels := dot_re.Split(name, -1)
+	if len(labels) < 2 {
+		return fmt.Errorf("DNS name must not be a top level domain.")
+	}
+
+	for index, label := range labels {
+		// The final label is empty
+		if index+1 == len(labels) && len(label) < 1 {
+			return nil
+		}
+
+		if len(label) < 1 {
+			return fmt.Errorf("subdomains must not be empty")
+		}
+		s := []rune(label)
+		if len(s) > 63 {
+			return fmt.Errorf("subdomainss must not have more than 63 character: %d.", len(s))
```

Comment:
here and elsewhere- s/subdomainss/subdomains

Reply:
Thanks. Fixed

Reply:
Done.